cpp/src/arrow/type.cc

```diff
@@ -478,6 +479,47 @@ std::string DictionaryType::ToString() const {
   return ss.str();
 }
 
+Result<bool> DictionaryType::ComparableWithoutUnification(
+    std::vector<const Array*> arrays) {
+  if (arrays.size() == 0) {
+    return Status::Invalid("need at least one input array");
+  }
+
+  for (auto a : arrays) {
+    if (a->type_id() != Type::DICTIONARY) {
+      return Status::TypeError("input arrays must be dictionary arrays");
+    }
+  }
```

Comment:
If this is written as a two-argument function where the inputs are DictionaryArray then these checks aren't needed

Reply:
I was following the signature of `DictionaryType::Unify()`. Would you prefer the signature `DictionaryArray::CanCompareIndices(const DictionaryArray&)` or so?